src/borrow.rs

```diff
@@ -0,0 +1,586 @@
+//! Types to safely create references into NumPy arrays
+//!
+//! It is assumed that unchecked code - which includes unsafe Rust and Python - is validated by its author
+//! which together with the dynamic borrow checking performed by this crate ensures that
+//! safe Rust code cannot cause undefined behaviour by creating references into NumPy arrays.
+//!
+//! # Rationale
+//!
+//! Rust references require aliasing discipline to be maintained, i.e. there must always
+//! exist only a single mutable (aka exclusive) reference or multiple immutable (aka shared) references
+//! for each object, otherwise the program contains undefined behaviour.
+//!
+//! The aim of this module is to ensure that safe Rust code is unable to violate these requirements on its own.
```

Comment:
This sentence is very long. Can a couple of the punctuation marks be swapped for full stops, just to ease it up for non-native readers?

Reply:
Split that up into three separate sentences.